Fix SQLite affinity type mapping to correctly find conversions

ajcvickers opened the pull request:
Fixes #16111

The issue was that when finding a type mapping based on type name affinity we were not checking that it was compatible with the CLR type. This resulted in short-circuiting of the conversion selection and hence the wrong mapping.

bricelam commented on test/EFCore.Sqlite.Tests/Storage/SqliteTypeMappingTest.cs:
```diff
@@ -34,22 +96,22 @@ public override void Create_and_clone_with_converter(Type mappingType, Type clrT
+        [InlineData("decimal(10,4)", null)]
+        [InlineData("boolean", null)]
+        [InlineData("unknown_type", null)]
+        [InlineData("", null)]
```
Let's also check what happens in RevEng with these column types now.

AndriySvyryd commented:
Should rename the branch to AffinityGauntlet

ajcvickers commented:
New version up: 
* Added scaffolding type mapper tests for SQLite
* Fixed scaffolding `byte[]` for empty and unknown store types
* Added affinity for `BIN` - @bricelam maybe not needed because `byte[]` is the default, but feels cleaner. Thoughts?
